shared/cancel/canceller.go

```diff
@@ -2,20 +2,60 @@ package cancel
 
 import (
 	"context"
+	"sync"
+	"time"
 )
 
-// Canceller is a simple wrapper for a cancellable context which makes the associated context.CancelFunc more easily
-// accessible.
+// Canceller behaves similarly to context.WithCancel(), but provides an embedded Cancel function, which makes it
+// more convenient to store in a long-lived struct, and does not use go routines.
+// It implements the context.Context interface.
```

Comment:
Suggest `but provides and embedded Cancel function` -> `but provides a Cancel function`, since the Cancel function is not embedded - it's defined on the `Canceller` struct.

Reply:
what word would you suggest that is different to the cancel function that a context provides?

Reply:
going for built-in

Reply:
Ok, how about:

> Canceller behaves similarly to context.WithCancel(), except that the cancel function is defined on Canceller itself, rather than being returned anonymously. This makes it more convenient to store in a long-lived struct. Additionally, Canceller does not use go routines. It implements the context.Context interface but Value and Deadline are not implemented.